fms/modules/tp.py

```diff
@@ -31,54 +25,229 @@ def setup_tp(self, rank: int, world_size: int) -> None:
         self.rank = rank
         self.world_size = world_size
 
-    def colwise_param_names(self) -> List[str]:
-        return []
+    def __get_tp_slices(
+        self,
+        tensor_value,
+        output_size_per_partition,
+        max_partition_sizes,
+    ):
+        cusum_max_partition_sizes = [0]
+        min_partition_size = min(max_partition_sizes)
+        for m in max_partition_sizes:
+            cusum_max_partition_sizes.append(
+                cusum_max_partition_sizes[-1] + (m // min_partition_size)
+            )
 
-    def rowwise_param_names(self) -> List[str]:
-        return []
+        for weight_i, replication in enumerate(max_partition_sizes):
+            yield slice(
+                (self.rank // max(1, self.world_size // replication))
+                * (
+                    output_size_per_partition
+                    * (max_partition_sizes[weight_i] // min_partition_size)
+                )
+                + (
+                    cusum_max_partition_sizes[weight_i]
+                    * tensor_value.shape[0]
+                    // cusum_max_partition_sizes[-1]
+                ),
+                ((self.rank // max(1, self.world_size // replication)) + 1)
+                * (
+                    output_size_per_partition
+                    * (max_partition_sizes[weight_i] // min_partition_size)
+                )
+                + (
+                    cusum_max_partition_sizes[weight_i]
+                    * tensor_value.shape[0]
+                    // cusum_max_partition_sizes[-1]
+                ),
+            )
```

Comment:
looks like we have 20 LOC stuffed into one line here. we should break this up and use intermediate variables if needed so it's easier to read&understand